README
PS1

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.

INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc PS1

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=PS1

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/PS1

    CPAN Ratings
        http://cpanratings.perl.org/d/PS1

    Search CPAN
        http://search.cpan.org/dist/PS1/

    Source Code
        git://github.com/ivanwills/PS1.git

COPYRIGHT AND LICENCE

Copyright (C) 2010 Ivan Wills

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for App-PS1

0.03   2016-09-11
        Updating for release of version 0.03
        Fixed spelling (Ivan Wills)
        Added documentation about status (Ivan Wills)
        Changed status to be shorter if nothing to report (Ivan Wills)
        Fixed missing 11 (Ivan Wills)
        Adding git status details to prompt (Ivan Wills)
        Added some description to A::P::P::Face (Ivan Wills)
        Added vtide files (Ivan Wills)

0.02   2016-01-01
        Updating for release of version 0.02
        Updating license (Ivan Wills)
        Made directory length configurable (and set to 30) (Ivan Wills)

0.01   2015-08-16
        Updating for release of version 0.01
        Removed creation of makefile (Ivan Wills)

0.009   2015-08-03
        Updating for release of version 0.009
        Fixed sizes (Ivan Wills)
        Fixing display (Ivan Wills)

0.008   2015-08-02
        Updating for release of version 0.008
        Fixed display of elements (Ivan Wills)

0.007   2015-08-01
        Updating for release of version 0.007
        Added documentation about new settings (Ivan Wills)
        Added logic to turn on and off file, dir and size (Ivan Wills)

0.006   2015-07-31
        Updating for release of version 0.006
        Added missing documentation (Ivan Wills)
        Updated dependencies (Ivan Wills)
        Converted options to using JSON (as documented) (Ivan Wills)
        Added getting options for future use (Ivan Wills)

0.005   2015-07-30
        Updating for release of version 0.005
        Added abreviation option for directories (Ivan Wills)
        Update TODO (Ivan Wills)

0.004   2015-07-01
        Updating for release of version 0.004
        Added env variable documentation into more places (Ivan Wills)

0.003   2015-06-16
        Updating for release of version 0.003
        Added documentation about environment variables (Ivan Wills)
        Added more to the description (Ivan Wills)
        Added image examples (Ivan Wills)
        Fixed param name (Ivan Wills)
        Better documentation for command (Ivan Wills)

0.002   2015-06-08
        Updating for release of version 0.002
        Added limiting dir name length (Ivan Wills)
        Set max branch name length to 60 characters (Ivan Wills)

0.001   2014-08-29
        Removed unused modules and cleaned up version number (Ivan Wills)
        Updated build (Ivan Wills)
        Updated tests to my current standards (Ivan Wills)
        Updated manifest (Ivan Wills)
        Added support for nvm's version of node switching (Ivan Wills)
        Fixed variable error (Ivan Wills)
        Converted to use subtests rather than just diag the sub-test results (Ivan Wills)
        Added config for travis-ci (Ivan Wills)
        Cleaned up POD to pass tests (Ivan Wills)
        Updated to latest version (Ivan Wills)
        Fixed readme (Ivan Wills)
        Added referenced commands to be ignored (Ivan Wills)
        Updating README description and changing to POD (Ivan Wills)
        Added extra logic to try to work out the original branch when in detached head state (Ivan Wills)
        Added blue style (Ivan Wills)
        Stopped warnings (Ivan Wills)
        Fixed version environment config (Ivan Wills)
        Made check more specific to stop error when there is a .git directory which isn't a git repository (Ivan Wills)
        Added eval to protect against plugins erroring (Ivan Wills)
        Updated manifest (Ivan Wills)
        Fixed not updating variable name (Ivan Wills)
        New plugin to display ruby version (Ivan Wills)
        Fixed abstracts to match spec per http://neilb.org/2014/02/19/compliant-abstract.html (Ivan Wills)
        Fixed dependency (Ivan Wills)
        Fixed colouring (Ivan Wills)
        Merge branch 'master' of github.com:ivanwills/App-PS1 (Ivan Wills)
        Fixed up documentation (Ivan Wills)
        Added showing perlbrew and nave node versions (Ivan Wills)
        Adding tag files (Ivan Wills)
        Fixed spelling (Ivan Wills)
        Added method documentation (Ivan Wills)
        Cleaned up boiler plate text (Ivan Wills)
        Started documenting methods (Ivan Wills)
        Removed non-existant files (Ivan Wills)
        Updated package files (Ivan Wills)
        Added setting the theme from the environment (Ivan Wills)
        Added green theme (Ivan Wills)
        Removed warn (Ivan Wills)
        Converted colours to use themes (Ivan Wills)
        Passing more things on (Ivan Wills)
        Some rationalisation of plugins and cleanup (Ivan Wills)
        Changed it so that the background colour can be configured by an environment variable (Ivan Wills)
        Removed colon (Ivan Wills)
        Added space (Ivan Wills)
        Added all recommended modules (Ivan Wills)
        Added configure_requires section (Ivan Wills)
        Merge branch 'master' of https://github.com/ivanwills/App-PS1 (Ivan Wills)
        Updated generated files (Ivan Wills)
        Stopped Perl::Critic complaining about stuff (Ivan Wills)
        Updated stopwords (Ivan Wills)
        Corrected copyright (Ivan Wills)
        Updated license text (Ivan Wills)
        Fixed copyright (Ivan Wills)
        Added ignoring generated files (Ivan Wills)
        Added ignoring generated files (Ivan Wills)
        Fixed copyright (Ivan Wills)
        Added missing dependencies (Ivan Wills)
        Started working out how to write a deamon for the prompt (Ivan Wills)
        Updated MANIFEST files (Ivan Wills)
        Automated update (Ivan Wills)
        Added place name words to be ignored by spelling (Ivan Wills)
        Fixed incorrect mobule name (Ivan Wills)
        initial commit (Ivan Wills)


lib/App/PS1.pm
package App::PS1;

# Created on: 2011-06-21 09:47:36
# Create by:  Ivan Wills
# $Id$
# $Revision$, $HeadURL$, $Date$
# $Revision$, $Source$, $Date$

use strict;
use warnings;
use Carp qw/cluck/;
use Data::Dumper qw/Dumper/;
use English qw/ -no_match_vars /;
use Term::ANSIColor;
use base qw/Class::Accessor::Fast/;

eval { require Term::Colour256 };
my $t256 = !$EVAL_ERROR;

our $VERSION = 0.03;

__PACKAGE__->mk_accessors(qw/ ps1 cols plugins bw low exit parts safe theme verbose/);

my %theme = (
    default => {
        # name          Low Colour  Hi Colour
        background   => [ 'black' , 'on_52'  ],
        marker       => [ 'black' , 246      ],
        up_time      => [ 'yellow', 'yellow' ],
        up_label     => [ 'black' , 'black'  ],
        branch       => [ 'cyan'  , 'cyan'   ],
        branch_label => [ 'black' , 'black'  ],
        date         => [ 'red'   , 'red'    ],
        face_happy   => [ 'green' , 46       ],
        face_sad     => [ 'red'   , 202      ],
        dir_name     => [ 'blue'  , 'blue'   ],
        dir_label    => [ 'black' , 'black'  ],
        dir_size     => [ 'cyan'  , 'cyan'   ],
    },
    green => {
        # name          Low Colour  Hi Colour
        background   => [ 'on_green', 'on_22'  ],
        marker       => [ 'black'   , 246      ],
        up_time      => [ 'yellow'  , 'yellow' ],
        up_label     => [ 'black'   , 'black'  ],
        branch       => [ 'white'   , 190      ],
        branch_label => [ 'black'   , 'black'  ],
        date         => [ 'red'     , 9        ],
        face_happy   => [ 'green'   , 46       ],
        face_sad     => [ 'red'     , 202      ],
        dir_name     => [ 'blue'    , 21       ],
        dir_label    => [ 'black'   , 'black'  ],
        dir_size     => [ 'cyan'    , 33       ],
    },
    blue => {
        # name          Low Colour  Hi Colour
        background   => [ 'on_blue' , 'on_30'  ],
        marker       => [ 'black'   , 236      ],
        up_time      => [ 'yellow'  , 'yellow' ],
        up_label     => [ 'black'   , 'black'  ],
        branch       => [ 'white'   , 190      ],
        branch_label => [ 'black'   , 'black'  ],
        date         => [ 'red'     , 52       ],
        face_happy   => [ 'green'   , 46       ],
        face_sad     => [ 'red'     , 52       ],
        dir_name     => [ 'blue'    , 21       ],
        dir_label    => [ 'black'   , 'black'  ],
        dir_size     => [ 'green'   , 46       ],
    },
);

sub new {
    my ($class, $params) = @_;
    my $self = $class->SUPER::new($params);

    $self->safe( $ENV{UNICODE_UNSAFE} ) if !defined $self->safe;
    $self->theme("default")             if !defined $self->theme;

    $theme{ $self->theme } ||= {};
    for my $name ( keys %{ $theme{ $self->theme } } ) {
        my $env = $ENV{ 'APP_PS1_' . uc $name };
        if ($env) {
            $theme{ $self->theme }{$name} = [ ( $env ) x 2 ];
        }
    }

    return $self;
}

sub sum(@) { ## no critic
    my $i = 0;
    $i += $_ || 0 for (@_);
    return $i;
}

sub cmd_prompt {
    my ($self) = @_;
    my $out = '';
    $self->parts([]);

    for my $param ( split /;/, $self->ps1 ) {
        my ( $plugin, $options ) = split /(?=[{])/, $param;
        next if $plugin !~ /^[a-z]+$/;
        next if !$self->load($plugin);

        $options = $self->parse_options($options, $plugin);
        my ($text, $size) = eval { $self->$plugin($options) };

        if ($size) {
            push @{$self->parts}, [ $text, $size ];
        }
    }

    my $total = $self->parts_size;
    my $spare = $self->cols - $total;
    my $spare_size = $spare / ( @{$self->parts} - 1 );

    while ($spare < 0 || $spare_size < 0) {
        pop @{$self->parts};
        if ( @{$self->parts} == 1 ) {
            $total = $self->parts_size;
            $spare = $self->cols - $total;
            $spare_size = $spare;
            last;
        }
        $total = $self->parts_size;
        $spare = $self->cols - $total;
        $spare_size = ( @{$self->parts} - 1 ) ? $spare / ( @{$self->parts} - 1 ) : 0;
    }

    if ( $total <= $self->cols ) {
        my $line = '';
        my $extra = 0;
        for my $i ( 0 .. @{$self->parts} - 2 ) {
            my $div_first  = $i ? 2 : 1;
            my $div_second = $i == @{$self->parts} - 2 ? 1 : 2;
            my $spaces;
            if ( $total < $self->cols / 2 ) {
                $spaces = ( $self->cols / ( @{$self->parts} - 1 ) - $self->parts->[$i][0] / $div_first - $self->parts->[$i + 1][0] / $div_second );
            }
            else {
                $spaces = $spare_size;
                $spare -= $spare_size - ( $spaces - int $spaces );
            }
            $extra += $spaces - int $spaces;

            $line .= $self->parts->[$i][1];
            $line .= ' ' x $spaces;
            if ( $extra > 1 ) {
                $line .= ' ' x $extra;
                $spare -= int $extra;
                $extra = $extra - int $extra;
            }
        }
        if ( $extra > 0.1 ) {
            $line .= ' ';
        }
        $line .= $self->parts->[-1][1];

        my $colour = $ENV{APP_PS1_BACKGROUND} || 52;
        $out = $self->colour('background') . $line . "\e[0m\n";
    }

    return $out;
}

sub parse_options {
    my ($self, $options_txt, $name) = @_;

    return {} if !$options_txt;

    require JSON::XS;

    my $options = eval { JSON::XS::decode_json($options_txt) };
    my $error   = $@;

    if ($error && $self->verbose) {
        cluck "Error reading $name\'s options ($options_txt)! $error\n";
    }

    return $options || {};
}

sub parts_size {
    my ($self) = @_;
    return sum map { $_->[0] } @{$self->parts};
}

sub load {
    my ($self, $plugin) = @_;

    $self->plugins({}) if !$self->plugins;

    return 1 if $self->plugins->{$plugin};

    my $module = 'App::PS1::Plugin::' . ucfirst $plugin;
    my $file   = 'App/PS1/Plugin/' . ( ucfirst $plugin ) . '.pm';
    eval { require $file };
    warn $@ if $@;
    return 0 if $@;

    push @App::PS1::ISA, $module;

    return $self->plugins->{$plugin} = 1;
}

sub surround {
    my ($self, $count, $text) = @_;

    return if !defined $text;

    my $left  = $self->safe ? '≺' : '<';
    my $right = $self->safe ? '≻' : '>';

    $count += 2;
    $text = $self->colour('marker') . "$left$text" . $self->colour('marker') . $right;
    return ($count, $text);
}

sub colour {
    my ($self, $name) = @_;
    my $colour = $theme{$self->theme}{$name} || [];
    return
          $self->bw || !$colour ? ''
        : $t256 && !$self->low  ? Term::Colour256::color($colour->[1])
        :                         Term::ANSIColor::color($colour->[0]);
}

1;

__END__

=head1 NAME

App::PS1 - Module to load PS1 status line elements

=head1 VERSION

This documentation refers to App::PS1 version 0.03.

=head1 SYNOPSIS

   # in your ~/.bashrc file
   export APP_PS1='face;branch;date;direcory;perl;node;ruby;uptime'
   export PS1="\[\`app-ps1 -e\$?\`\]\n\u@\h \\\$ "

=head1 DESCRIPTION

This is the engine for the C<app-ps1> command.

=head1 SUBROUTINES/METHODS

=head3 C<new ( $param_hash )>

Param: C<ps1>   Str  What plugins to show on the prompt
Param: C<low>   Bool Use low (16 bit colour)
Param: C<bw>    Bool Don't use any colour (black and white)
Param: C<theme> Str  Use colour theme
Param: C<exit>  Int  The last program's exit code
Param: C<cols>  Int  The number of columns wide to assume the terminal is

Return: App::PS1 - A new object

Description:

=head3 C<sum ( @list )>

Adds the values in list and returns the result.

=head3 C<cmd_prompt ()>

Display the command prompt

=head3 C<parts_size ()>

calculate the size of the prompt parts

=head3 C<load ()>

Load plugins

=head3 C<surround ()>

Surround the text with brackets

=head3 C<colour ($name)>

Get the theme colour for C<$name>

=head3 C<parse_options ($options)>

Parses the JSON $options txt.

=head1 DIAGNOSTICS

=head1 CONFIGURATION AND ENVIRONMENT

Lots of environment variables are used to configure the command prompt

=over 4

=item C<$APP_PS1>

Sets the elements to be displayed (overridden by C<--ps1>)

Default 'face;branch;date;directory;uptime',

=item C<$APP_PS1_THEME>

Sets the colour theme for the prompt

=over 4

=item *

default

=item *

green

=item *

blue

=back

Default 'default',

=item C<$PS1_COLS>

If L<Term::Size::Any> is not installed you can configure the width of your
screen by setting this parameter.

Default 90,

=item C<$UNICODE_UNSAFE>

If set to a true value this will allow UTF8 characters to be used displaying
the prompt

Default not set

=item C<$APP_PS1_BACKGROUND>

Set the line's background colour

Default 52

=back

=head1 DEPENDENCIES

=head1 INCOMPATIBILITIES

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module.

Please report problems to Ivan Wills (ivan.wills@gmail.com).

Patches are welcome.

=head1 AUTHOR

Ivan Wills - (ivan.wills@gmail.com)

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2011 Ivan Wills (14 Mullion Close, Hornsby Heights, NSW, Australia 2077)
All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.  This program is
distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

=cut
